packages/edit-post/src/components/header/mode-switcher/index.js

```diff
@@ -55,12 +55,32 @@ function ModeSwitcher() {
 		return null;
 	}
 
-	if ( ! isRichEditingEnabled || ! isCodeEditingEnabled ) {
-		return null;
+	let selectedMode = mode;
+	if ( ! isRichEditingEnabled && mode === 'visual' ) {
+		selectedMode = 'text';
+	}
+	if ( ! isCodeEditingEnabled && mode === 'text' ) {
+		selectedMode = 'visual';
 	}
 
 	const choices = MODES.map( ( choice ) => {
-		if ( choice.value !== mode ) {
+		if ( ! isCodeEditingEnabled && choice.value === 'text' ) {
+			choice = {
+				...choice,
+				disabled: true,
+				info: __( 'Code editor is disabled for this site.' ),
+			};
+		}
+		if ( ! isRichEditingEnabled && choice.value === 'visual' ) {
+			choice = {
+				...choice,
+				disabled: true,
+				info: __(
+					'Enable the visual editor in your profile settings.'
```

Comment:
Minor language nit: in some languages, those strings sound a little harsh, almost like forcing the user to do something. 

As a less demanding alternative I'd suggest:

```suggestion
					'You can enable the visual editor in your profile settings.'
```